### PR TITLE
Add authorization caching support

### PR DESCRIPTION
**Pull request by periklis**
Based on the [discussion](https://github.com/observatorium/opa-openshift/pull/1#discussion_r671218810) this PR adds caching support for authorization queries against the k8s/openshift API server. The basic tenet here is to protect the api server from pulling the same authz response on each request.

### CODE REVIEW

**Review by bill3tt — COMMENTED** [review]
Reading @squat 's comment
>  It would be considerably less load to use a Kubernetes client informer to watch for all of the namespaces and store them in a local cache, which can later be queried on every request.

This PR implements our own caching, rather than using the Kubernetes client informer as suggested - is that deliberate? Is that because we need to pass along the subjects auth token?

Also - it is not clear to me why we need both in-memory and memcache clients? My preference would be to [KISS](https://en.wikipedia.org/wiki/KISS_principle), especially if we don't explicitly need one of the options (probably memcache?) - it can be added when needed.

**Comment by periklis** [comment]
> Reading @squat 's comment
> 
> > It would be considerably less load to use a Kubernetes client informer to watch for all of the namespaces and store them in a local cache, which can later be queried on every request.
> 
> This PR implements our own caching, rather than using the Kubernetes client informer as suggested - is that deliberate? Is that because we need to pass along the subjects auth token?

The PR implements a cache for authorization `DataResponseV1` results per auth token for two particular reasons instead of a cache on the informer level:
1. There are two different clients for two api servers in openshift. One for the k8s apiserver (used for the `SelfSubjectAccessReviews`) and one for the openshift apiserver (used to list accessible projects). This would require two individual caches for a single token with all the burden to evict in parallel.
2. Talking to both k8s/openshift clients is combined in a single authz response. If we would cache on the client level, this adds the complexity to evict hiearchically, i.e. if an an initial `SelfSubjectAccessReview` returns `status.allowed: true` is cached along with a list of accessible projects, we need on a subsequent `status.allowed:false` for the same auth token evict first the cache holding the `SelfSubjectAccessReview` response and in lock-step the other cache holding the list of projects.

In summary maintaining two client-level caches would introduce some extra premium on eviction with all the locking required to do this safely. Thus, the PR introduces a high level cache on the OPA DataResponseV1 level.

> 
> Also - it is not clear to me why we need both in-memory and memcache clients? My preference would be to [KISS](https://en.wikipedia.org/wiki/KISS_principle), especially if we don't explicitly need one of the options (probably memcache?) - it can be added when needed.

The `opa-openshift` service can and will be used in different scenarios, where the availability of a `memcached` instance is not given. In addition the apiserver on k8s/openshift is not a typical HA-proofed component, scaling horizontally indefinitely. An HA-proofed apiserver always holds a single replica answering queries and a set of (2-4 replicas) in stand-by to take over if the first replica fails. The worst thing one can do is to push the cluster into constant leader-election because each replica is flooded in sequence with requests and gives up. Thus the `opa-openshift` should work also self-sufficiently to not push unnecessary to the api server. One scenario in particular it will be used is for in-cluster Loki instances w/o `memcached`.

File: internal/config/config.go [hunk 1]
```diff
@@ -61,6 +68,11 @@ func ParseFlags() (*Config, error) {
 	flag.StringVar(&cfg.Opa.Rule, "opa.rule", "allow", "The name of the OPA rule for which opa-openshift should provide a result, see https://www.openpolicyagent.org/docs/latest/policy-language/#rules.") //nolint:lll
 	flag.StringVar(&cfg.Opa.Matcher, "opa.matcher", "", "The label key of the OPA label matcher returned to the requesting client.")
 
+	// Memcached flags
+	flag.StringSliceVar(&cfg.Memcached.Servers, "memcached", nil, "One or more Memcached server addresses.")
```
Comment (ianbillett) [inline internal/config/config.go:72]:
```suggestion
	flag.StringSliceVar(&cfg.Memcached.Servers, "memcached", nil, "One or more Memcached server addresses. If this flag is unset, in-memory cache will be used.")
```

**Review by bill3tt — APPROVED** [review]
I'm afraid I don't know the code well enough to pore over every detail - only qualm would be that the caching code paths are not explicitly unit-tested.

LGTM? :shrug:

**Review by bwplotka — APPROVED** [review]
Thanks! LGTM, just minor nits not blocking.

**Pull request merged**